training/splitting_dataset.py

```diff
@@ -0,0 +1,132 @@
+import json
+from collections import defaultdict
+from typing import Dict, List, Tuple
+
+import pandas as pd
+from sklearn.model_selection import train_test_split
+
+from training.analysis_functions import DataFrameType
+
+
+_DEFAULT_TEST_SIZE: float = 0.25
+"""Default value of fraction of the dataset to include in test split."""
+
+
+def _get_book_names_from_dataframe(dataframe: DataFrameType) -> List[str]:
+    """
+    Collects unique values of "book" column in dataframe.\n
+    :param dataframe: Processed dataframe.
+    :return: List of books' names.
+    """
+    try:
+        return dataframe.book.unique()
+    except AttributeError as e:
+        print(f"_get_book_names_from_dataframe: Error occurred: {e}.")
+        return []
+
+
+def _get_grouped_book_names(dataframe: pd.DataFrame) -> Dict[int, List[str]]:
+    """
+    Groups names of books present in "book" column in dataframe by the year the books were published.\n
+    :param dataframe: Processed dataframe.
+    :return: Dictionary containing list of books for each year.
+    """
+    with open("../training/book_year_mapping.json", "r") as mapping_file:
+        book_year_mapping = json.loads("".join(mapping_file.readlines()))
+
+    grouped_books_per_year = defaultdict(list)
+    book_names = _get_book_names_from_dataframe(dataframe)
+    for book_name in book_names:
+        if book_year := book_year_mapping.get(book_name):
+            grouped_books_per_year[book_year].append(book_name)
+        else:
+            grouped_books_per_year[0].append(book_name)
+    return grouped_books_per_year
+
+
+def _get_dataframe_with_oldest_books(
+    dataframe: DataFrameType, test_size: float = _DEFAULT_TEST_SIZE
+) -> DataFrameType:
+    """
+    Extracts from dataframe rows with the oldest books. Number of extracted rows is equal to
+    a fraction of initial number of rows defined in test_size.\n
+    :param dataframe: Processed dataframe.
+    :param test_size: Fraction of the dataset to include in test split. It should be a float number between 0.0 and 1.0.
+    :return: Dataframe containing extracted rows.
+    """
+    books_per_year = _get_grouped_book_names(dataframe)
+    filtered_dataframe = pd.DataFrame(columns=dataframe.columns)
+    remaining_rows_num = int((1 - test_size) * dataframe.shape[0])
+    for year in sorted(books_per_year.keys()):
+        for book_name in books_per_year[year]:
+            book_df = dataframe.loc[dataframe["book"] == book_name]
```

Comment:
Jeżeli mielibyśmy informację o dokładnym dniu wydania książki, to wystarczyłoby posortować malejąco po dacie wydania i iterować po tym